### PR TITLE
auth/cert: Add certificate extensions as metadata

### PR DESCRIPTION
**Pull request by peterverraedt**
This adds certificate extensions as metadata to be used in vault policy templates. We parse the ASN1 objectidentifiers to a string separated by dashes instead of dots, to ease the use in golang templating language.

Fixes: #13347

Refs: #10503 

Signed-off-by: Peter Verraedt <peter.verraedt@kuleuven.be>

### CODE REVIEW

**Comment by pmmukh** [comment]
Hey @peterverraedt ! Sorry for the delay here, I've been caught up with other stuff but I'll be sure to review this PR this week, and thanks for the prompt response to the feedback we gave!

File: builtin/credential/cert/backend_test.go [hunk 1]
```diff
@@ -1652,6 +1691,7 @@ func testAccStepCert(
 			"allowed_uri_sans":             testData.uris,
 			"allowed_organizational_units": testData.organizational_units,
 			"required_extensions":          testData.ext,
+			"metadata_extensions":          testData.metadata_ext,
```
Comment (pmmukh) [inline builtin/credential/cert/backend_test.go:1694]: Would it make sense to rename this to something like `allowed_metadata_extensions` ? I think having the field name reflect the purpose, that extensions with these oids are allowed to be added into identity metadata, like allowed_uri_sans / allowed_organizational_units, might be clearer ?

File: builtin/credential/cert/path_certs.go [hunk 2]
```diff
@@ -114,6 +114,15 @@ formatted as "oid:value". Expects the extension value to be some type of ASN1 en
 All values much match. Supports globbing on "value".`,
 			},
 
+			"metadata_extensions": {
+				Type: framework.TypeCommaStringSlice,
+				Description: `A comma-separated string or array of extensions
+formatted as "oid" or "oid:alias". Upon successfull authentication, these extensions will
```
Comment (pmmukh) [inline builtin/credential/cert/path_certs.go:120]: Couple points here.
1. Do we need the 2 forms, oid and oid:alias ? As in, if we just have a comma separate list of oids, and always use the santiized ( dots to dashes ) version of it as the metadata key, would there be an issue ? I'm guessing redabale metadata keys is the argument there, but wanted to check.
2. If we do want the oid:alias form, I'd recommend using a different word and not `alias`, as that is a loaded word, being a key concept in the identity system.

File: builtin/credential/cert/path_login.go [hunk 3]
```diff
@@ -409,6 +419,47 @@ func (b *backend) matchesCertificateExtensions(clientCert *x509.Certificate, con
...
+	// If no metadata extensions are configured, return an empty map
+	if len(config.Entry.MetadataExtensions) == 0 {
+		return map[string]string{}
+	}
+
+	// Build a map of the extensions that should be set as metadata
+	oidMap := map[string]string{}
+	for _, metadataExt := range config.Entry.MetadataExtensions {
+		parts := strings.SplitN(metadataExt, ":", 2)
+		if len(parts) == 1 {
```
Comment (pmmukh) [inline builtin/credential/cert/path_login.go:434]: A comment here explaining the conditional might be nice imo

File: builtin/credential/cert/path_login.go [hunk 4]
```diff
@@ -409,6 +419,47 @@ func (b *backend) matchesCertificateExtensions(clientCert *x509.Certificate, con
...
+			// Avoid dots in metadata keys - hard to work with in policy templates.
+			oidMap[metadataExt] = strings.ReplaceAll(metadataExt, ".", "-")
+
```
Comment (pmmukh) [inline builtin/credential/cert/path_login.go:437]:
```suggestion
```

File: builtin/credential/cert/path_login.go [hunk 5]
```diff
@@ -409,6 +419,47 @@ func (b *backend) matchesCertificateExtensions(clientCert *x509.Certificate, con
...
+			continue
+		}
+
+		oidMap[parts[0]] = parts[1]
+	}
+
+	metadata := make(map[string]string, len(config.Entry.MetadataExtensions))
+	for _, ext := range clientCert.Extensions {
+		for oidString, alias := range oidMap {
```
Comment (pmmukh) [inline builtin/credential/cert/path_login.go:446]: Do we need to loop through the map in here with the string checking, or could we do something like
`if alias, ok:= oidMap[ext.IdString()]; ok {` ?
The suggestion is less for a time complexity improvement, which I think doesn't matter tbh, and more for readability.

File: builtin/credential/cert/path_login.go [hunk 6]
```diff
@@ -89,19 +89,29 @@ func (b *backend) pathLogin(ctx context.Context, req *logical.Request, data *fra
...
+	metadata := map[string]string{
+		"cert_name":        matched.Entry.Name,
+		"common_name":      clientCerts[0].Subject.CommonName,
+		"serial_number":    clientCerts[0].SerialNumber.String(),
+		"subject_key_id":   certutil.GetHexFormatted(clientCerts[0].SubjectKeyId, ":"),
+		"authority_key_id": certutil.GetHexFormatted(clientCerts[0].AuthorityKeyId, ":"),
+	}
+
+	for k, v := range b.certificateExtensionsMetadata(clientCerts[0], matched) {
+		if _, ok := metadata[k]; ok {
```
Comment (pmmukh) [inline builtin/credential/cert/path_login.go:101]: Could you add a comment explaining this conditional too please ?

**Review by pmmukh — COMMENTED** [review]
Some comments / questions, overall the PR looks pretty great! :)

File: builtin/credential/cert/backend_test.go (hunk 1 above)
Comment (peterverraedt) [reply builtin/credential/cert/backend_test.go:1694]: I wanted to initially avoid `allowed_metadata_extensions` as the other `allowed_*` keys restrict access to the authentication endpoint for certificates lacking those properties. But I follow you, and it indeed makes sense to put `allowed_metadata_extensions`, as it is a filter on the oids that we'll accept to set as metadata.

File: builtin/credential/cert/path_certs.go (hunk 2 above)
Comment (peterverraedt) [reply builtin/credential/cert/path_certs.go:120]: We don't need aliases apart from readability wherever the metadata is used, and actually putting the dashes-separated oid as metadata key will make it clear throughout that the metadata originated as certificate extensions. As we're renaming the configuration value to `allowed_metadata_extensions`, it also makes more sense to just accept a list of oids. I'll remove the aliasing.

File: changelog/13348.txt [hunk 7]
```diff
@@ -0,0 +1,3 @@
+```release-note:feature
```
Comment (pmmukh) [inline changelog/13348.txt:1]: I think this should be `improvement` not `feature`, as `feature` usually refers to large changes in a major release.

File: builtin/credential/cert/path_certs.go [hunk 8]
```diff
@@ -114,6 +114,14 @@ formatted as "oid:value". Expects the extension value to be some type of ASN1 en
 All values much match. Supports globbing on "value".`,
 			},
 
+			"allowed_metadata_extensions": {
+				Type: framework.TypeCommaStringSlice,
+				Description: `A comma-separated string or array of oid extensions.
```
Comment (pmmukh) [inline builtin/credential/cert/path_certs.go:119]: Does a comma separated string work here ? I didn't see a split by comma way of parsing the input in the code, hence figured i'd check.

**Review by pmmukh — APPROVED** [review]
couple minor nits, olgtm!

**Comment by pmmukh** [comment]
Also, could you please rebase this branch off main ? There's some test failures that should be fixed in main now.

File: builtin/credential/cert/path_certs.go (hunk 8 above)
Comment (pmmukh) [reply builtin/credential/cert/path_certs.go:119]: ugh i missed the `TypeCommaStringSlice` there, please ignore!

**Pull request merged**